hubblestack/loader.py

```diff
@@ -0,0 +1,1292 @@
+                    break
+        if mod_name in self.loaded_modules:
+            return self.loaded_modules[mod_name]
+        else:
+            raise AttributeError(mod_name)
+
+    def missing_fun_string(self, function_name):
+        '''
+        Return the error string for a missing function.
+
+        This can range from "not available' to "__virtual__" returned False
+        '''
+        mod_name = function_name.split('.')[0]
+        if mod_name in self.loaded_modules:
+            return '\'{0}\' is not available.'.format(function_name)
+        else:
+            try:
+                reason = self.missing_modules[mod_name]
+            except KeyError:
+                return '\'{0}\' is not available.'.format(function_name)
+            else:
+                if reason is not None:
+                    return '\'{0}\' __virtual__ returned False: {1}'.format(mod_name, reason)
+                else:
+                    return '\'{0}\' __virtual__ returned False'.format(mod_name)
+
+    def _refresh_file_mapping(self):
+        '''
+        refresh the mapping of the FS on disk
+        '''
+        # map of suffix to description for imp
+        if self.opts.get('cython_enable', True) is True:
+            try:
+                global pyximport
+                pyximport = __import__('pyximport')  # pylint: disable=import-error
+                pyximport.install()
+                # add to suffix_map so file_mapping will pick it up
+                self.suffix_map['.pyx'] = tuple()
+            except ImportError:
+                log.info('Cython is enabled in the options but not present '
+                    'in the system path. Skipping Cython modules.')
+        # Allow for zipimport of modules
+        if self.opts.get('enable_zip_modules', True) is True:
+            self.suffix_map['.zip'] = tuple()
+        # allow for module dirs
+        self.suffix_map[''] = ('', '', MODULE_KIND_PKG_DIRECTORY)
+
+        # create mapping of filename (without suffix) to (path, suffix)
+        # The files are added in order of priority, so order *must* be retained.
+        self.file_mapping = salt.utils.odict.OrderedDict()
+
+        opt_match = []
+
+        def _replace_pre_ext(obj):
+            '''
+            Hack so we can get the optimization level that we replaced (if
+            any) out of the re.sub call below. We use a list here because
+            it is a persistent data structure that we will be able to
+            access after re.sub is called.
+            '''
+            opt_match.append(obj)
+            return ''
+
+        for mod_dir in self.module_dirs:
+            try:
+                # Make sure we have a sorted listdir in order to have
+                # expectable override results
+                files = sorted(
+                    x for x in os.listdir(mod_dir) if x != '__pycache__'
+                )
+            except OSError:
+                continue  # Next mod_dir
+            if six.PY3:
```

Comment:
Since we are on Python-3 now. We can actually straight away use statements for PY3, and can remove code branches for PY2

Reply:
I wanted to get this done sooner than later so I elected to skip removing six everywhere. I removed some. I think you're right though. I'll remove the rest today.

Reply:
removed six